fitbenchmarking/parsing/horace_parser.py

```diff
@@ -0,0 +1,194 @@
+"""
+This file implements a parser for Horace problem sets.
+"""
+import os
+import typing
+
+import matlab
+import numpy as np
+
+from fitbenchmarking.parsing.fitbenchmark_parser import FitbenchmarkParser
+from fitbenchmarking.utils.exceptions import ParsingError
+from fitbenchmarking.utils.matlab_engine import ENG as eng
+from fitbenchmarking.utils.matlab_engine import add_persistent_matlab_var
+
+
+def horace_on():
+    """
+    Turning Horace and SpinW on in matlab
+    """
+    if "HORACE_LOCATION" in os.environ and "SPINW_LOCATION" in os.environ:
+        horace_location = os.environ["HORACE_LOCATION"]
+        spinw_location = os.environ["SPINW_LOCATION"]
+        eng.evalc("restoredefaultpath")
+        eng.evalc(f"addpath('{horace_location}')")
+        eng.evalc("horace_on")
+        eng.evalc(f"addpath('{spinw_location}')")
+        eng.evalc("spinw_on")
+    elif "HORACE_LOCATION" not in os.environ and \
+         "SPINW_LOCATION" in os.environ:
+        raise ParsingError('Could not parse SpinW problem. Please ensure '
+                           'that HORACE_LOCATION is specfied as a environment '
+                           'variable')
+    elif "HORACE_LOCATION" in os.environ and \
+         "SPINW_LOCATION" not in os.environ:
+
+        raise ParsingError('Could not parse SpinW problem. Please ensure '
+                           'that SPINW_LOCATION is specfied as a environment '
+                           'variable')
+    else:
+        eng.evalc("horace_on")
```

Comment:
```suggestion
    eng.evalc("restoredefaultpath")

    horace_location = os.environ.get("HORACE_LOCATION", None)
    if horace_location:
        eng.evalc(f"addpath('{horace_location}')")

    try:
        eng.evalc("horace_on")
    except matlab.engine.MatlabExecutionError:
        raise ParsingError('Could not parse SpinW problem. Please ensure '
                           'that HORACE_LOCATION ("{horace_location}") is specfied as a environment '
                           'variable and contains "horace_on.m"')

    spinw_location = os.environ.get("SPINW_LOCATION", None)
    if spinw_location:
        eng.evalc(f"addpath('{spinw_location}')")

    try:
        eng.evalc("spinw_on")
    except matlab.engine.MatlabExecutionError:
        raise ParsingError('Could not parse SpinW problem. Please ensure '
                           'that SPINW_LOCATION ("{spinw_location}") is specfied as a environment '
                           'variable and contains "spinw_on.m"')

```

Reply:
This implementation is not going to work. I specifically wrote function so you don't need to set the `horace_location` and the `spinw_location` if you have a normal installation of horace. This is mention in the docs. Unless you would prefer that the environment variables are required ? I have checked this locally and it doesn't work.

Reply:
Yeah, I think it might not work because I left the extraneous `restoredefaultpath` in.

Reply:
I am not confident enough to put this in. This causes too much confusion with how to set up `spinw` as `spinw_on` is not a global variable locally(my pc) but it's global variable on IDAaaS (spinw problems work on both). I might just remove `spinw_on`  as it is not needed to run spinw problems and it is not mention in the SpinW documentation. The path is still needed (https://www.spinw.org/installation/). This would fix the the current error messages, however the  `restoredefaultpath` is needed for IDAaaS. The whole point of this logical is for it to work on both IDAaaS and locally seamlessly. Opinions ?

Reply:
Ideally this will not be needed for IDAaaS (though admittedly it currently is), but I need to work out what I want to ask them for with regards to not loading default Horace on startup. `spinw_on` adds the required paths, that's its job.

Reply:
We are adding the required spinw paths manually anyways so `spinw_on` can just be removed then.   I'll probably keep the current implementation for now and create another issue suggesting that the `horace_on` function should to be changed to the code below once the  IDAaaS issue has been resolved.

```
    horace_location = os.environ.get("HORACE_LOCATION", None)
    if horace_location:
        eng.evalc(f"addpath('{horace_location}')")
    try:
        eng.evalc("horace_on")
    except matlab.engine.MatlabExecutionError:
        raise ParsingError('Could not parse SpinW problem. Please ensure '
                           f'that HORACE_LOCATION ("{horace_location}") is'
                           ' specfied as a environment '
                           'variable and contains "horace_on.m"')
    spinw_location = os.environ.get("SPINW_LOCATION", None)
    if spinw_location:
        eng.evalc(f"addpath('{spinw_location}')")
```

Reply:
Sounds like a reasonable solution. The main reason for the suggestion was just that I think it's easier to parse what's intended and I suspect `else: horace_on` might catch people by surprise